man/bmon.8
.TH "bmon" "8" "" "Bandwidth Monitor" "bmon"
.SH "NAME"
bmon \- bandwidth monitor and rate estimator
.SH "SYNOPSIS"
.B bmon
[\fB\-\-show\-all\fR]
[\fB\-\-use\-si\fR]
[\fB\-\-input\fR=\fIMODULE\fR]
[\fB\-\-output\fR=\fIMODULE\fR]
[OPTIONS...]

.SH "DESCRIPTION"
bmon is a monitoring and debugging tool to capture networking related
statistics and prepare them visually in a human friendly way. It
features various output methods including an interactive curses user
interface and a programmable text output for scripting.

.SH "OPTIONS"
.PP
\fB \-h\fR, \fB\-\-help\fR
.RS 4
Prints a short help text and exits\&.
.RE
.PP
\fB \-V\fR, \fB\-\-version\fR
.RS 4
Prints the versioning identifier and exits\&.
.RE
.PP
\fB \-i\fR, \fB\-\-input=\fRMODULE[:OPTIONS][,MODULE...]
.RS 4
Set list of input modules to load and use. Multiple modules can be used
in parallel. bmon automatically loads a useful and working input module
by default. See INPUT MODULES for more details.
.RE
.PP
\fB \-o\fR, \fB\-\-output=\fRMODULE[:OPTIONS][,MODULE...]
.RS 4
Set list of output modules to load and use. Multiple modules can be used
in parallel. By default, bmon will use the curses output mode, if that is
not available due to an incompatible console it will fall back to a simple
text mode. See OUTPUT MODULES for more details.
.RE
.PP
\fB \-U\fR, \fB\-\-use\-si\fR
.RS 4
Use SI unit system (1KB = 1'000 bytes) instead of 1KB = 1'024 bytes.
.RE
.PP
\fB \-f\fR, \fB\-\-configfile=\fRFILE
.RS 4
Set alternative path to configuration file.
.RE
.PP
\fB \-p\fR, \fB\-\-policy=\fRPOLICY
.RS 4
Set policy defining which network interfaces to display. See
INTERFACE SELECTION for more details.
.RE
.PP
\fB \-a\fR, \fB\-\-show\-all\fR
.RS 4
Display all interfaces, even interface that are administratively down.
.RE
.PP
\fB \-r\fR, \fB\-\-read\-interval=\fRFLOAT
.RS 4
Set interval in seconds in which input modules read statistics from their
source. The default is 1.0 seconds.
.RE
.PP
\fB \-R\fR, \fB\-\-rate\-interval=\fRFLOAT
.RS 4
Set interval in seconds in which the rate per counter is calculated.
The default is 1.0 seconds.
.RE
.PP
\fB \-b\fR, \fB\-\-use\-bit\fR
.RS 4
Show rates in bits per second instead of bytes per second.
.RE
.PP
\fB \-L\fR, \fB\-\-lifetime=\fRFLOAT
.RS 4
Set lifetime of an element in seconds before it is no longer displayed
without receiving any statistical updates. The default is 30 seconds.
.RE

.SH "INPUT MODULES"
.PP
Input modules provide statistical data about elements. Each element consists
of attributes which represents a counter, a rate, or a percentage. Elements
may carry additional child elements to represent a hierarchy. Each element is
assigned to a group defined by the input module. Input modules are polled in
the frequence of the configured read interval.
.PP
The following input modules are available:
.TP
\fBnetlink\fR
Uses the Netlink protocol to collect interface and traffic control statistics
from the kernel. This is the default input module.

.TP
\fBproc\fR
Reads interface statistics from the /proc/net/dev file. This is considered a
legacy interface and provided for backwards compatibily reasons. This is a
fallback module if the Netlink interface is not available.

.TP
\fBdummy\fR
Programmable input module for debugging and testing purposes.

.TP
\fBnull\fR
No data collected.

.PP
To receive additional information about a module, run the module with the
"help" option set like this:

.PP
.RS 4
bmon \-i netlink:help
.RE

See MODULE CONFIGURATION for more details.

.SH "OUTPUT MODULES"
.PP
Output modules display or export the statistical data collected by input
modules. Multiple output modules can be run at the same time. bmon will
not prevent possible conflicts such as multiple output modules writing to
the console.
.PP
The following output modules exist:

.TP
\fBcurses\fR
Interactive curses based text user interface providing real time rate
estimations and a graphical representation of each attribute. Press '?'
to display the quick reference guide. This is the default output mode.

.TP
\fBascii\fR
Simple programmable text output intended for human consumption. Capable
of printing list of interfaces, detailed counters and graphs to the
console. This is the default fallback output mode if curses is not
available.

.TP
\fBformat\fR
Fully scriptable output mode inteded for consumption by other programs.
See the module help text for additional information.

.TP
\fBnull\fR
Disable output.

.PP
To receive additional information about a module, run the module with the
"help" option set like this:

.PP
.RS 4
bmon \-o curses:help
.RE

See MODULE CONFIGURATION for more details.

.SH "MODULE CONFIGURATION"
.PP
The syntax to configure modules is as follows:
.PP
.RS 4
ARGUMENT ::= mod1[:OPTS][,mod2[:OPTS]...]
.br
OPTS     ::= OPTION[;OPTION...]
.br
OPTION   ::= option[=value]
.RE

.PP
Run the module with option "help" to receive the list of options for each
module:

.PP
.RS 4
bmon \-i module:help
.RE

.SH "INTERFACE SELECTION"
.PP
The following syntax is used to define the interface selection policy:
.PP
.RS 4
SELECTION ::= NAME[,NAME[,...]]
.br
NAME      ::= [!]interface
.RE

.PP
The interface name may contain the character '*' which will act as a wildcard
and represents any number of any character type, e.g. eth*, h*0, ...

.PP
Examples:

.PP
.RS 4
lo,eth0,eth1
.br
eth*,!eth0
.RE

.SH "EXAMPLES"
.PP
To run bmon in curses mode monitoring the interfaces eth0
and eth1:
.PP
.RS 4
\fBbmon \-p eth0,eth1 \-o curses\fP
.RE
.PP
To run bmon in format mode, monitoring any eth* interfaces, with a specified
format string:
.PP
.RS 4
\fBbmon \-p \(aqeth*\(aq \-o format:fmt=\(aq$(element:name) $(attr:rxrate:packets)\en\(aq\fP
.RE
.PP

.SH "FILES"
/etc/bmon.conf
.br
$HOME/.bmonrc

.SH "SEE ALSO"
.PP
\fBip\fR(8),
\fBnetstat\fR(8),
\fBifconfig\fR(8),
\fBnetlink\fR(7),

.SH "AUTHOR"
Thomas Graf <tgraf@suug.ch> among others


NEWS
HEAD

v4.0 - Dec 13, 2016
 * Use monotonic clock instead of realtime clock
 * Pick default selected interface based on policy
 * Collect RX NoHandler statistics if available (Linux)
 * CentOS installation instructions
 * Proper stdout flush in ASCII mode
 * Bugfixes

v3.9 - Jul 19, 2016
 * Color support
 * Add ability to reset statistics from curses UI
 * NetBSD compile fix
 * Option to enable info display by default
 * Additional IPv6 statistics
 * Various fixes

v3.8 - July 25, 2015
 * Don't disable Netlink if TC stats are unavailable

v3.7 - November 22, 2014
 * Bugfixes
 * Documentation updates
 * Provide minimal interface information on BSD
 
v3.6 - November 22, 2014
 * Build fix for uclinux
 * Fix LICENSE links

v3.5 - August 30, 2014
 * Fixes for all defects identified by coverity
 * Fix accuracy issue on total rate calculation
 * Travis-CI support
 * Various other small bugfixes

v3.4 - August 24, 2014
 * Bugfixes
   * blank screen with config file
   * quick-help toggle with '?' in curses
 * Better bmon.conf example

v3.3 - July 6, 2014
 * MacOS X port
 * Only initialize curses module if actually used
 * Assorted bug and spelling fixes
 * Various build fixes


src/bmon.c
/*
 * src/bmon.c		   Bandwidth Monitor
 *
 * Copyright (c) 2001-2013 Thomas Graf <tgraf@suug.ch>
 * Copyright (c) 2013 Red Hat, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include <bmon/bmon.h>
#include <bmon/conf.h>
#include <bmon/attr.h>
#include <bmon/utils.h>
#include <bmon/input.h>
#include <bmon/output.h>
#include <bmon/module.h>
#include <bmon/group.h>

int start_time;

struct reader_timing rtiming;

static char *usage_text =
"Usage: bmon [OPTION]...\n" \
"\n" \
"Options:\n" \
"Startup:\n" \
"   -i, --input=MODPARM             Input module(s)\n" \
"   -o, --output=MODPARM            Output module(s)\n" \
"   -f, --configfile=PATH           Alternative path to configuration file\n" \
"   -h, --help                      Show this help text\n" \
"   -V, --version                   Show version\n" \
"\n" \
"Input:\n" \
"   -p, --policy=POLICY             Element display policy (see below)\n" \
"   -a, --show-all                  Show all elements (even disabled elements)\n" \
"   -r, --read-interval=FLOAT       Read interval in seconds (float)\n" \
"   -R, --rate-interval=FLOAT       Rate interval in seconds (float)\n" \
"   -s, --sleep-interval=FLOAT      Sleep time in seconds (float)\n" \
"   -L, --lifetime=LIFETIME         Lifetime of an element in seconds (float)\n" \
"\n" \
"Output:\n" \
"   -U, --use-si                    Use SI units\n" \
"   -b, --use-bit                   Display in bits instead of bytes\n" \
"\n" \
"Module configuration:\n" \
"   modparm := MODULE:optlist,MODULE:optlist,...\n" \
"   optlist := option;option;...\n" \
"   option  := TYPE[=VALUE]\n" \
"\n" \
"   Examples:\n" \
"       -o curses:ngraph=2\n" \
"       -o list            # Shows a list of available modules\n" \
"       -o curses:help     # Shows a help text for curses module\n" \
"\n" \
"Interface selection:\n" \
"   policy  := [!]simple_regexp,[!]simple_regexp,...\n" \
"\n" \
"   Example: -p 'eth*,lo*,!eth1'\n" \
"\n" \
"Please see the bmon(8) man pages for full documentation.\n";

static void do_shutdown(void)
{
	static int done;
	
	if (!done) {
		done = 1;
		module_shutdown();
	}
}

static void sig_exit(void)
{
	do_shutdown();
}

void quit(const char *fmt, ...)
{
	static int done;
	va_list args;

	va_start(args, fmt);
	vfprintf(stderr, fmt, args);
	va_end(args);

	if (!done) {
		done = 1;
		do_shutdown();
	}

	exit(1);
}

static inline void print_version(void)
{
	printf("bmon %s\n", PACKAGE_VERSION);
	printf("Copyright (C) 2001-2015 by Thomas Graf <tgraf@suug.ch>\n");
	printf("Copyright (C) 2013 Red Hat, Inc.\n");
	printf("bmon comes with ABSOLUTELY NO WARRANTY. This is free " \
	       "software, and you\nare welcome to redistribute it under " \
	       "certain conditions. See the source\ncode for details.\n");
}

static void parse_args_pre(int argc, char *argv[])
{
	DBG("Parsing arguments pre state");

	for (;;)
	{
		char *gostr = "+:hvVf:";

		struct option long_opts[] = {
			{"help", 0, NULL, 'h'},
			{"version", 0, NULL, 'v'},
			{"configfile", 1, NULL, 'f'},
			{NULL, 0, NULL, 0},
		};
		int c = getopt_long(argc, argv, gostr, long_opts, NULL);
		if (c == -1)
			break;
		
		switch (c)
		{
			case 'f':
				set_configfile(optarg);
				break;

			case 'h':
				print_version();
				printf("\n%s", usage_text);
				exit(1);
			case 'V':
			case 'v':
				print_version();
				exit(0);
		}
	}
}

static int parse_args_post(int argc, char *argv[])
{
	DBG("Parsing arguments post state");

	optind = 1;

	for (;;)
	{
		char *gostr = "i:o:p:r:R:s:aUb" \
			      "L:hvVf:";

		struct option long_opts[] = {
			{"input", 1, NULL, 'i'},
			{"output", 1, NULL, 'o'},
			{"policy", 1, NULL, 'p'},
			{"read-interval", 1, NULL, 'r'},
			{"rate-interval", 1, NULL, 'R'},
			{"sleep-interval", 1, NULL, 's'},
			{"show-all", 0, NULL, 'a'},
			{"use-si", 0, NULL, 'U'},
			{"use-bit", 0, NULL, 'b'},
			{"lifetime", 1, NULL, 'L'},
			{NULL, 0, NULL, 0},
		};
		int c = getopt_long(argc, argv, gostr, long_opts, NULL);
		if (c == -1)
			break;
		
		switch (c)
		{
			case 'i':
				if (input_set(optarg))
					return 1;
				break;

			case 'o':
				if (output_set(optarg))
					return 1;
				break;

			case 'p':
				cfg_setstr(cfg, "policy", optarg);
				break;

			case 'r':
				cfg_setfloat(cfg, "read_interval", strtod(optarg, NULL));
				break;

			case 'R':
				cfg_setfloat(cfg, "rate_interval", strtod(optarg, NULL));
				break;

			case 's':
				cfg_setint(cfg, "sleep_time", strtoul(optarg, NULL, 0));
				break;

			case 'a':
				cfg_setbool(cfg, "show_all", cfg_true);
				break;

			case 'U':
				cfg_setbool(cfg, "use_si", cfg_true);
				break;

			case 'b':
				cfg_setbool(cfg, "use_bit", cfg_true);
				break;

			case 'L':
				cfg_setint(cfg, "lifetime", strtoul(optarg, NULL, 0));
				break;

			case 'f':
				/* Already handled in pre getopt loop */
				break;

			default:
				quit("Aborting...\n");
				break;

		}
	}

	return 0;
}

#if 0
static void calc_variance(timestamp_t *c, timestamp_t *ri)
{
	float v = (ts_to_float(c) / ts_to_float(ri)) * 100.0f;

	rtiming.rt_variance.v_error = v;
	rtiming.rt_variance.v_total += v;

	if (v > rtiming.rt_variance.v_max)
		rtiming.rt_variance.v_max = v;

	if (v < rtiming.rt_variance.v_min)
		rtiming.rt_variance.v_min = v;
}
#endif

int main(int argc, char *argv[])
{
	unsigned long sleep_time;
	double read_interval;
	
	start_time = time(NULL);
	memset(&rtiming, 0, sizeof(rtiming));
	rtiming.rt_variance.v_min = FLT_MAX;

	/*
	 * Early initialization before reading config
	 */
	conf_init_pre();
	parse_args_pre(argc, argv);

	/*
	 * Reading the configuration file */
	configfile_read();

	/*
	 * Late initialization after reading config
	 */
	if (parse_args_post(argc, argv))
		return 1;
	conf_init_post();

	module_init();

	read_interval = cfg_read_interval;
	sleep_time = cfg_getint(cfg, "sleep_time");

	if (((double) sleep_time / 1000000.0f) > read_interval)
		sleep_time = (unsigned long) (read_interval * 1000000.0f);

	DBG("Entering mainloop...");

	do {
		/*
		 * E  := Elapsed time
		 * NR := Next Read
		 * LR := Last Read
		 * RI := Read Interval
		 * ST := Sleep Time
		 * C  := Correction
		 */
		timestamp_t e, ri, tmp;
		unsigned long st;

		float_to_timestamp(&ri, read_interval);

		/*
		 * NR := NOW
		 */
		update_timestamp(&rtiming.rt_next_read);
		
		for (;;) {
			output_pre();

			/*
			 * E := NOW
			 */
			update_timestamp(&e);

			/*
			 * IF NR <= E THEN
			 */
			if (timestamp_le(&rtiming.rt_next_read, &e)) {
				timestamp_t c;

				/*
				 * C :=  (NR - E)
				 */
				timestamp_sub(&c, &rtiming.rt_next_read, &e);

				//calc_variance(&c, &ri);

				/*
				 * LR := E
				 */
				copy_timestamp(&rtiming.rt_last_read, &e);

				/*
				 * NR := E + RI + C
				 */
				timestamp_add(&rtiming.rt_next_read, &e, &ri);
				timestamp_add(&rtiming.rt_next_read,
				       &rtiming.rt_next_read, &c);


				reset_update_flags();
				input_read();
				free_unused_elements();
				output_draw();
				output_post();
			}

			/*
			 * ST := Configured ST
			 */
			st = sleep_time;

			/*
			 * IF (NR - E) < ST THEN
			 */
			timestamp_sub(&tmp, &rtiming.rt_next_read, &e);

			if (tmp.tv_sec < 0)
				continue;

			if (tmp.tv_sec == 0 && tmp.tv_usec < st) {
				if (tmp.tv_usec < 0)
					continue;
				/*
				 * ST := (NR - E)
				 */
				st = tmp.tv_usec;
			}
			
			/*
			 * SLEEP(ST)
			 */
			usleep(st);
		}
	} while (0);

	return 0; /* buddha says i'll never be reached */
}

static void __init bmon_init(void)
{
	atexit(&sig_exit);
}
